torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp

```diff
@@ -2202,15 +2203,30 @@ c10::intrusive_ptr<ProcessGroupNCCL::WorkNCCL> ProcessGroupNCCL::initWork(
       desyncDebug_,
       enableTiming_.load(),
       dist_debug_level_);
-  r->trace_id_ = NCCLTraceBuffer::get()->record(
-      uid_,
-      seq_,
-      // create a string copy of profilingTitle
-      profilingTitle ? profilingTitle : "",
-      inputs,
-      outputs,
-      r->ncclStartEvent_.get(),
-      r->ncclEndEvent_.get());
+  if (record) {
+    // Ideally record every work that we enqueue, rather than every work we
+    // create.
+    // - at the time of this PR we do not currently enqueue every created work
+    // - but it is unsafe to steal refs to start/end cuda events from Works that
+    // may go
+    //   out of scope before flight recorder has retired them,
+    //   so we must ensure that any work that is initialized via initWork will
+    //   be enqueued
```

Comment:
Do we guarantee "any work that is initialized via initWork will be enqueued'? Since I see a few calls of " initWork(device, rank_, optype, "nccl:coalesced", {}, {}, enqueue);", when  'enqueue' is false, work is inited but will not be enqueued

Reply:
Maybe it is clearer if we can logically separate the concept of 'record' from 'work'.   we can call NCCLTraceBuffer::get()->record anytime, e.g, outside of initWork, when ever we want to add a flight record, and it might or might not be to a 'Work'

Reply:
we can change the design to decouple record() from work, but that is a bigger scope than this PR.  This PR attempts to just make the current design more explicit.

The complexity comes from (1) record takes a ref to cuda events owned by work and lifecycle managed by work, (2) work lifecycle managed by enqueue and watchdog loop, (3) flight recorder 'retire' ends lifecycle after enqueued work is procesesed in watchdog, (4) i think @zdevito intentionally avoided making new cuda events for recorder vs work, for perf reasons

So, with the current design we should ensure any time a record happens, its work is kept alive.  But i did not add a 'mechanism' for this, just wrote out the rules.  Open to suggestions for how to do this.

Reply:
Agree that we should not making new cuda events. But I saw sometimes record is already happen outside of initwork intentionally, we can make this style consistent?
<img width="732" alt="image" src="https://github.com/pytorch/pytorch/assets/4711970/86af2dff-f960-4a15-8aa6-04ee3ca079ea">

Reply:
Then, I think it is fine to keep this way for now, I just found it a bit hard to wrap my head on which case we should record with initWork and which cases we should not